Use Vector2i instead of Vector2 for Image get_pixelv and set_pixelv

aaronfranke opened the pull request:
Since Vector2i is now available, we should use it for APIs that currently accept two ints. These methods previously had Vector2 versions, so they were converted to Vector2i. The int versions are kept in parallel in this case.

akien-mga reviewed (approved):
Seems good to me. I'll double check with @reduz but I think this is indeed the intended direction.

Xrayez commented:
What's the performance cost? I mostly prefer `get_pixel(x, y)` over `get_pixelv()`. If you look at the changes in this PR, the use case for `get_pixel(x, y)` is actually easy and efficient traversal of pixels within the image (or image rect).

I'd prefer that we still have both `get_pixel(x: int, y: int)` and `get_pixelv(pos: Vector2i)` or any other combination of names, because I don't see a strong reason for removing the `get_pixel(x: int, y: int)`, both methods **complement** each other and serve different use cases...

akien-mga commented:
It's true that the code of this PR adds a lot of extra allocations of `Point2i(x, y)` in code that's otherwise still scalar based.

aaronfranke commented:
@Xrayez I think the performance cost is minimal, if any. These are allocated on the stack, so they should take up the same space in memory. In terms of efficiency with GDScript, I think this is actually more efficient because you only need a single Variant instead of two Variants (so that's 20 or 24 bytes instead of 40 or 48 bytes).

There are indeed a lot of instances in the codebase where `(x, y)` is shorter. I'm not against keeping a version with two ints, but if it's kept then it should be named so that the Vector2i version is displayed as the primary option (I'm thinking `get_pixel` for Vector2i and `get_pixeli` for two ints).

Xrayez commented:
I'd also say it's a quite heavy compatibility breakage which will impact a lot of existing C++ image processing stuff in Godot (via modules). As a user/developer I have to do a lot more changes than this PR introduces, just to achieve similar/compatible API in relation to Godot API. I'm not lazy but I don't think removing `(x: int, y: int)` version is really that worth it.

> In terms of efficiency with GDScript, I think this is actually more efficient because you only need a single Variant instead of two Variants (so that's 20 or 24 bytes instead of 40 or 48 bytes).

For GDScript this is likely OK, but the performance benefits can be achieved even without removing `(x, y)` method. 😛 

> I'm not against keeping a version with two ints, but if it's kept then it should be named so that the Vector2i version is displayed as the primary option (I'm thinking `get_pixel` for Vector2i and `get_pixeli` for two ints).

Well, oftentimes I wish we had proper method overloading in Godot in the first place...

Xrayez commented:
There are similar existing `*v`-like methods:
- [`Vector2.posmod/posmodv`](https://docs.godotengine.org/en/latest/classes/class_vector2.html#class-vector2-method-posmod)
- [`TileMap.set_cell/set_cellv`](https://docs.godotengine.org/en/latest/classes/class_tilemap.html#class-tilemap-method-set-cell)
- I've even proposed `RandomNumberGenerator.randv_circle()` in #43343 similarly.

So, I think this is a common thing to provide both integer and vector versions in API, I don't think that those are somehow confusing to beginners either, and mostly provides convenience, if not talking about performance, I'd just make sure that documentation is clear regarding which version is considered as the "main" one.

aaronfranke commented:
@Xrayez `posmodv` is different, since the non-vector version takes a single parameter instead of two. For `TileMap`, I made this PR awhile ago: #39976. GridMap also used to be the same way, but I made a PR for that too and that PR was merged.

Xrayez commented:
Yeah as you may figure, I wasn't happy with #39958 and future #39976, lets see what users say once Godot 4.0 stable is there.

But yeah, I guess I cannot go against development philosophy behind Godot (whatever it is). 😛

aaronfranke commented:
@Xrayez In the absence of a strong argument against this, the default choice should be "only vectors". Otherwise we would have to provide 2 versions of tons of different APIs. For example, if we can [translate by a Vector2](https://docs.godotengine.org/en/latest/classes/class_transform2d.html#class-transform2d-method-translated), what about translation by two floats? Providing methods with multiple int or float arguments only makes sense as an exceptions to the general rule of "only vectors".

If there is a specific case in which providing multiple arguments instead of a vector is a large improvement, then that can be discussed. Maybe Image is one such case, truthfully I haven't used this API much, so I'd say it's best to ask @reduz. But I don't think this applies to TileMap/GridMap. In any case, having Vector2i as the primary option seems best to me.

Xrayez commented:
IMO, unlike "translate by a Vector2" case, the difference between those is exactly "for loop" traversal (also applies to `TileMap` and similar):
```gdscript
for y in height:
    for x in width:
        image.set_pixel(x, y, Color())
```

With this PR:
```gdscript
for y in height:
    for x in width:
        image.set_pixel(Vector2(x, y), Color())
```

If we go this route, I'd like that we implement vector literals as well: godotengine/godot-proposals#374, so this would just become:

```gdscript
for y in height:
    for x in width:
        image.set_pixel((x, y), Color())
```

Or again, implement method overloading. 🙂

Xrayez commented:
I've done some benchmarking with doctest in https://github.com/Xrayez/godot/tree/pixelvi-perf-test:

Command used to run tests:
```
godot --test --source-file="*test_image*" --test-case="*stress*" --duration
```

# With this PR

## `Point2`, 2048x2048 image fill with `set_pixel`

![godot_image_fill_vector](https://user-images.githubusercontent.com/17108460/99524546-3fb17500-29a1-11eb-85e4-36417d677e55.png)

## `Point2`, 4096x4096image fill with `set_pixel`

![godot_image_fill_vector_4096](https://user-images.githubusercontent.com/17108460/99524577-4c35cd80-29a1-11eb-8271-432457244e54.png)

# W/o changes in this PR

## `(x, y)`, 2048x2048 image fill with `set_pixel`

![godot_image_fill_xy](https://user-images.githubusercontent.com/17108460/99524660-6cfe2300-29a1-11eb-82bf-fbdd1b97dd9f.png)

## `(x, y)`, 4096x4096 image fill with `set_pixel`

![godot_image_fill_xy_4096](https://user-images.githubusercontent.com/17108460/99524688-79827b80-29a1-11eb-8ee2-832f615a4b32.png)

MSVC compiler.

# Conclusion

On average, this PR decreases performance by up to **30-40%** on smaller images, and up to **20-30%** on larger images.

I suggest:
1. Revert core changes in this PR.
2. Rename `set/get_pixel` to `set/get_pixeli`.
3. Rename `set/get_pixelv` to `set/get_pixel`.

This way, you can achieve desired consistency/cleanness of API, *and* allow both C++ and GDScript to be performant.

But as I said earlier, these changes are for the sake of changes, there's no absolute reason why these kind of changes have to be done in the first place.

I've been using `Image` API extensively almost like on "day 1" using Godot, so the performance aspect is quite important to me, and `Image` is already on the edge performance-wise. It means that some optimization stuff I've done to save me like 20% of performance while processing images will be thrown out with this PR... 😕

aaronfranke commented:
@Xrayez Your test case is flawed, but I didn't realize this until after I updated this PR. Basically, you are using `Point2` instead of `Point2i`, so right now it's converting the arguments to floats, then putting them in a `Point2`, then converting that to a `Point2i` which has to convert the floats back to ints, then passing that to the method. As opposed to just putting the ints in a `Point2i` and passing that without any conversions with floats and `Point2`. I would appreciate if you could re-test this with a fixed test case :slightly_smiling_face: 

Anyway, here's what I was going to say before I realized the above (so the below may no longer be valid):

> Due to the severity of the performance difference, I think it would make sense to keep the version with two ints indicated as the primary option by having no suffix. I updated this PR, removing a lot of the diffs, so that now it only does the following:
> 
> * Change `get_pixelv`/`set_pixelv` to use `Vector2i` instead of `Vector2`.
> 
> * Use the `v` version in a few places where vector math does actually make sense, like when adding to an existing vector.
>
> * Update the documentation for the above.

Xrayez commented:
@aaronfranke no, I've tested the worst case (with using `Point2` over `Point2i`), but didn't mention this explicitly (read on).

Here's a result with `Point2i`:

![image](https://user-images.githubusercontent.com/17108460/99669332-912c3380-2a77-11eb-90f0-9d1eb5f84a1e.png)

I wanted to prove another point but waited until you could find the bottleneck yourself: it's *very* easy to misuse the API if you're not aware of integer based version of `Point2i`, and as in your case, to actually find the root cause of a performance bottleneck. The core API should provide computational robustness.

That said, I never got it faster than `0.24` seconds as with results w/o changes in this PR. But yeah the integer version of `Point2` works a bit faster, so now we're talking about **10-20%** decrease in performance on average, which is still quite crucial if you ask me.

> Due to the severity of the performance difference

Sorry but I do not accept any performance loss using Image API and I'd interpret that as a regression myself, why make something perform worse, especially when it's easy to retain the current API?

Xrayez commented:
It's also quite unusual to see `set/get_pixel()` to use a struct for specifying coordinates. There are quite a bunch of image processing libraries written in C which also suggest (x, y) usage, so these changes add a layer of complexity and lack of compatibility with existing image processing libraries (and even algorithms).

https://pike.lysator.liu.se/generated/manual/modref/ex/predef_3A_3A/SDL/Surface/set_pixel.html
https://docs.microsoft.com/en-us/dotnet/api/system.drawing.bitmap.setpixel?view=dotnet-plat-ext-5.0
https://www.partow.net/programming/bitmap/documentation/classbitmap__image.html
https://www.kite.com/python/docs/salt.modules.sensehat.set_pixel
https://microbit-micropython.readthedocs.io/en/v1.0.1/display.html

If someone can find an image processing library or API which uses a struct for pixel coordinates, please post some links as well.

Xrayez commented on core/io/image.cpp:
```diff
@@ -2567,15 +2567,14 @@ void Image::blit_rect_mask(const Ref<Image> &p_src, const Ref<Image> &p_mask, co
 		for (int j = 0; j < dest_rect.size.x; j++) {
-			int src_x = clipped_src_rect.position.x + j;
-			int src_y = clipped_src_rect.position.y + i;
+			Point2i point = Point2i(j, i);
+			Point2i source = clipped_src_rect.position + point;
```
I wouldn't play with the devil here, just like with other changes in `Image`. This is ok for the rest of the engine where image is used if proved to be easier to read and whatnot, but again `int` would likely perform faster.

aaronfranke committed:
Use Vector2i instead of Vector2 for Image get_pixelv and set_pixelv

Co-authored-by: Andrii Doroshenko <xrayez@gmail.com>

aaronfranke commented:
@Xrayez Alright, I think you know best here. I changed this so that it's now very much playing it safe. Let me know what you think.

Xrayez reviewed (approved):
Ok, thanks for going through this. 😛

akien-mga commented:
Thanks to both of you for the constructive discussion and testing. I agree with the consensus :)

ArdaE commented:
@Xrayez, just out of curiosity, have you tried compiling with optimizations on? Your tests would be flawed if you complied with `scons tests=yes` as you mention above. Any decent compiler should be able to optimize away the seeming overhead that using Point2i would introduce.

Xrayez commented:
@ArdaE to be honest, I haven't, yet the issue would still be with API that is not commonly used for image processing.

> Any decent compiler should be able to optimize away the seeming overhead that using Point2i would introduce.

Exactly, not every compiler would be able to optimize, especially with so many platforms which Godot can be compiled for (including experimental [Raspberry Pi](https://www.reddit.com/r/godot/comments/af1cji/compiling_godot_31_for_raspberry_pi/) ports).

Also, AFAIK people do ship games with debug builds in Godot, to eliminate various release/debug inconsistencies.

ArdaE commented:
@Xrayez, I think pretty much every platform uses gcc (GNU), clang, or MSVC, which should all be able to optimize this case with ease. I'm not familiar with Raspberry Pi, but I'm guessing that too must be using either gcc or clang.

Both the original PR and the final PR rely on optimizations (the former for the manual conversions to/from `Point2i`, which can be optimized at compile time, and the latter for the inlining of the new Point2i-based API functions, which can only be done at link time). So I'm not sure if the "[don't] rely on compiler to optimize things" argument really applies here, as both solutions rely on optimizations. However, I see your point about most existing code in Godot being one way and handling that as the default case.

> AFAIK people do ship games with debug builds in Godot

Anyone who chooses to release a game using a Debug build cannot complain about poor performance. I don't think optimizing that case should be a concern or priority for Godot developers. People can use the `release_debug` build to help debug field issues and have the same safety checks that debug builds have if that's what's desired. That still might wreak havoc on the performance, but it wouldn't be nearly as bad as releasing a full-on Debug build to end users.

Anyway, please don't take this as an argument for or against the versions of the functions you advocate for, as I don't care much either way. However, I was surprised by your argument that the original PR caused slow-downs, as that would be a serious concern for a lot of other similar code within Godot and my own projects. So I've run a couple of tests based on your performance test code (your original tests were too short to be accurate; I've repeated your set of for loops 100 times to reduce noise due to other processes in the system, hence the longer durations below).

**Windows - MSVC:**
```
===============================================================================
TEST CASE:  [Stress][Image] 1-Image fill by 2048 (x,y)
3.883582 s: [Stress][Image] 1-Image fill by 2048 (x,y)
===============================================================================
TEST CASE:  [Stress][Image] 2-Image fill by 2048 (Point2i)
3.841098 s: [Stress][Image] 2-Image fill by 2048 (Point2i)
===============================================================================
TEST CASE:  [Stress][Image] 3-Image fill by 4096 (x,y)
15.920441 s: [Stress][Image] 3-Image fill by 4096 (x,y)
===============================================================================
TEST CASE:  [Stress][Image] 4-Image fill by 4096 (Point2i)
15.856829 s: [Stress][Image] 4-Image fill by 4096 (Point2i)
===============================================================================
```
I don't have MinGW installed on my Windows computer, so I don't know how gcc would fair. The results from clang below on my slightly faster MacBook are similar though.

**macOS - clang:**
```
===============================================================================
TEST CASE:  [Stress][Image] 1-Image fill by 2048 (x,y)
3.253238 s: [Stress][Image] 1-Image fill by 2048 (x,y)
===============================================================================
TEST CASE:  [Stress][Image] 2-Image fill by 2048 (Point2i)
3.223625 s: [Stress][Image] 2-Image fill by 2048 (Point2i)
===============================================================================
TEST CASE:  [Stress][Image] 3-Image fill by 4096 (x,y)
12.903808 s: [Stress][Image] 3-Image fill by 4096 (x,y)
===============================================================================
TEST CASE:  [Stress][Image] 4-Image fill by 4096 (Point2i)
12.858390 s: [Stress][Image] 4-Image fill by 4096 (Point2i)
===============================================================================
```
Incidentally, the Point2i versions of the functions were marginally faster than the versions you advocate in both tests above, but that would be just noise. If one ran the test again and again, and in a different order, the results should converge at pretty much the same numbers. Even the relative placement of the code in memory can make a difference, which can change from build to build, so we can ignore minute differences. For all practical purposes, there's no difference between the "(x,y)" and "Point2i" versions.

By the way, these results are from a test that has the (x,y)-based `set_pixel` function along with the original PR's Point2i-based `set_pixel` function within a single build. Because the final PR that was modified based on your discussion requires link-time inlining of the Point2i-based `set_pixel`, it may be slower depending on the linker.


PS. Another thing that's not ideal about your test setup is that it runs within an editor/tool build of Godot. A more proper performance test would have been within a deployment build of Godot built in `release` mode, which cannot be used in editor/tool builds. I was able to reduce the difference by disabling the conditional debug code within `set_pixel` in the clang build (hadn't thought of that yet at the time I ran the MSVC test above).

PPS. I see that you're a Godot contributor; thank you for all your contributions!

Xrayez commented:
@ArdaE ok, thanks for taking the time to profile this. I'm not particularly versed with profiling techniques myself and I'm always learning (someone out there has always more experience to prove someone wrong 😃). But in my experience, even when you think that something will be indeed optimized by the compiler, it's not always the case (I understand this is anecdotal, but still...)

Also, debug builds tend to be much slower in general than release builds. If we apply this reasoning to everything, then we'll also have much slower development iterations of the engine/C++ custom modules side (which I'm highly fond of). And yeah, as I said, I do process images quite religiously (I also have a test suite which processes hundreds of images in debug builds). So yeah, I agree that this is not an issue to the end user, but this should be still somewhat important to developers.

I understand that the performance impact this PR would have in debug builds wouldn't cause major hiccups in most cases, but it's just too easy to neglect the performance aspect of development with Godot to think that "everything is going to be fixed for you anyway" approach, and this approach can eventually turn into development philosophy (yes, I'm digging that deep and look that far). So, consider my arguments to have ideological sense. 😄

That said, you mostly won't find me modifying core implementations as I don't allow myself to do those changes, especially if there's no *substantial* reason for those changes (who wants to introduce regressions? 🙂), but some contributors find the courage to do so. Due to this, please understand my anxious reactions regarding such changes.

> Anyone who chooses to release a game using a Debug build cannot complain about poor performance.

To clarify, I'm literally saying that the logic may be different in debug/release builds as to cause crashes in release builds: godotengine/godot-proposals#1589.

Xrayez commented:
> PS. Another thing that's not ideal about your test setup is that it runs within an editor/tool build of Godot. A more proper performance test would have been within a deployment build of Godot built in `release` mode, which cannot be used in editor/tool builds. I was able to reduce the difference by disabling the conditional debug code within `set_pixel` in the clang build (hadn't thought of that yet at the time I ran the MSVC test above).

Forgot to mention that there's no way to run tests in `release` (`tools=no`) builds currently. And tests are run separately from production code anyway (they currently even have different entry point), so the setup itself is quite different too (doesn't run main loop, nor initializes display):

https://github.com/godotengine/godot/blob/7e009a167110e79caa102c5b88363fd32b75293c/main/main.cpp#L453-L469

ArdaE commented:
> there's no way to run tests in `release` (`tools=no`) builds currently.

I wouldn't use the built-in test framework for performance tests. As you mention, there's no way to run them in a release build as is (unless one modifies that part of Godot first). My approach would have been to write a little module or modify an existing built-in function of Godot temporarily to test this (e.g. take any function you can trigger from GDScript, and change it so it runs your performance test instead of doing what it is supposed to do; then deploy a tiny project that triggers a call to that function).
